File: cppProjekt/FINAL PROJECT PO/Animal.cpp
#include "Animal.h"
#include "Turtle.h"

vector <bool> Animal::checkDirections() {
	bool left = true;
	bool right = true;
	bool up = true;
	bool down = true;
	vector <bool> directions;
	if ( getX() - 2 < 0) {
		left = false;
	}
	else if (getX() + 2 >  world->getX()) {
		right = false;
	}
	if ( getY() - 2 < 0) {
		up = false;
	}
	else if ( getY() + 1 > world->getY()) {
		down = false;
	}
	directions.push_back(up);
	directions.push_back(left);
	directions.push_back(right);
	directions.push_back(down);
	setOldPosition(this->getPozycja());
	return directions;

}

void Animal::Akcja() {
	/*
	bool left = true;
	bool right = true;
	bool up = true;
	bool down = true;
	if (START_X + getX() - 2 < START_X) {
		left = false;
	}
	else if (START_X + getX() + 2> START_X + world->getX()) {
		right = false;
	}
	if (START_Y + getY() - 2 < START_Y) {
		up = false;
	}
	else if (START_Y + getY() + 3> START_Y + world->getY()) {
		down = false;
	}
	*/
	/// <summary>
	/// tut
	/// </summary>
	
	
	vector <bool> directions = checkDirections();
	bool up = directions[0];
	bool left = directions[1];
	bool right = directions[2];
	bool down = directions[3];
	
	
	
	while (true) {
		int a = rand() % 100;
		

		if (a >= 0 && a < 25 && up == true) {
			SetPozycja(this->getX(), this->getY() - 1);
			break;
		}
		else if (a >= 25 && a < 50 && down == true) {
			SetPozycja(this->getX(), this->getY() + 1);
			break;
		}

		else if (a >= 50 && a < 75 && left == true) {
			SetPozycja(this->getX() - 1, this->getY());
			break;
		}
		else if (a >= 75 && right == true) {
			SetPozycja(this->getX() + 1, this->getY());
			break;
		}
		
	}
	if (world->getElementInMap(this->getPozycja()) != nullptr && world->getElementInMap(this->getPozycja())->getModel() != this->getModel()) {
		world->getElementInMap(this->getPozycja())->Kolizja(this);
	}
	else if(world->getElementInMap(this->getPozycja()) != nullptr && world->getElementInMap(this->getPozycja())->getModel() == this->getModel() && this->getAge() > 5 && world->getElementInMap(this->getPozycja())->getAge() > 5){
	
		Breeding(world->getElementInMap(this->getPozycja()));
		this->SetPozycja(getOldPosition().getX(), getOldPosition().getY());
	}
	this->plusAge();
}

void Animal::Kolizja(Organizm* other) {
	
	if (this->GetSila()>other->GetSila()) {
		world->deleteOrganizm(other);
		if (other->GetSila() > 0) {
			string result = this->getName()+ " killed " + other->getName() + " on position "+this->getPozycjaStr()+".";
			world->info.addInformation(result);
		}
	}
	else if (this->GetSila() < other->GetSila()) {

		string result = other->getName() + " killed " + this->getName() + " on position " + this->getPozycjaStr() + ".";
		world->info.addInformation(result);
		world->deleteOrganizm(this);
	}
	else if (this->getAge() > other->getAge()) {
			string result = this->getName() + " killed " + other->getName() + " on position " + this->getPozycjaStr() + ".";
			world->info.addInformation(result);
			world->deleteOrganizm(other);
		
	}
	else {
		string result = other->getName() + " killed " + this->getName() + " on position " + this->getPozycjaStr() + ".";
		world->info.addInformation(result);
		world->deleteOrganizm(this);
	}
	
}



Animal::Animal(TypOrganizmu typOrganizmu, World* world, Punkt pozycja, int sila, int inicjatywa)
	:Organizm(typOrganizmu, world, pozycja, sila, inicjatywa)
{
	int seed = std::chrono::duration_cast<std::chrono::milliseconds>(
		std::chrono::system_clock::now().time_since_epoch()
		).count();
	std::srand(seed);
	this->isAnimal = true;
}
void Animal::Breeding(Organizm* other) {
	vector <bool> directions = checkDirections();
	bool up = directions[0];
	bool left = directions[1];
	bool right = directions[2];
	bool down = directions[3];
	Punkt newPosition;
	if (world->getElementInMap({this->getX()+1, this->getY()}) == nullptr && right == true) {
		newPosition.setX(this->getX() + 1);
		newPosition.setY(this->getY());
	}
	else if (world->getElementInMap({ this->getX() - 1, this->getY() }) == nullptr && left == true) {
		newPosition.setX(this->getX() - 1);
		newPosition.setY(this->getY());
	}
	else if (world->getElementInMap({ this->getX() , this->getY()-1 }) == nullptr && up == true) {
		newPosition.setX(this->getX() );
		newPosition.setY(this->getY() -1);
	}
	else {
		newPosition.setX(this->getX());
		newPosition.setY(this->getY() + 1);
	}
	
	world->addOrganizm(createOrganizm(this->typOrganizmu, this->world, newPosition));
	string result = "A new " + this->getName() + " was born in the position ";
	world->info.addInformation(result);
}

File: javaProject/src/main/java/project_java/Animal.java
package project_java;

import java.awt.*;
import java.net.MalformedURLException;
import java.util.Random;

public abstract class Animal extends Organizm{
    //private static final Point position = ;

    public Animal(){}

    public Animal(World world,Point position, int age, int power, int init){
        super(world,position,age,power,init);
        isAnimal = true;
        oldPosition = position;
        //oldPosition = new Point(0,0);
    }
    public Animal(World world, Point position){
        super(world, position);
    }
    @Override
    public Color getColor() {
        return null;
    }

    @Override
    public void Akcja() {
        boolean left = true;
        boolean right = true;
        boolean up = true;
        boolean down = true;
        oldPosition = position;
        if (0+this.getPosition().getX() - 1 < 0) {
            left = false;
        }
        else if (0 + this.getPosition().getX() + 2> 0 + world.getX()) {
            right = false;
        }
        if (this.getPosition().getY() - 1 < 0) {
            up = false;
        }
        else if (this.getPosition().getY()+ 2> 0 + world.getY()) {
            down = false;
        }

        Random random = new Random();
        while (true) {



            int a = random.nextInt(100);

            if (a >= 0 && a < 25 && up == true) {
                position.changePositionUp();

                break;
            }
            else if (a >= 25 && a < 50 && down == true) {
               position.changePositionDown();
                break;
            }

            else if (a >= 50 && a < 75 && left == true) {
                position.changePositionLeft();
                break;
            }
            else if (a >= 75 && right == true) {
                position.changePositionRight();
                break;
            }

        }
        if (world.whoIs(this.getPosition(), this) != null && world.whoIs(this.getPosition(), this).getSign() != this.getSign()) {
            world.whoIs(this.getPosition(), this).Kolizja(this);
        }
        else if(world.whoIs(this.getPosition(), this) != null && world.whoIs(this.getPosition(), this).getSign() == this.getSign() && this.getAge() > 10 && world.whoIs(this.getPosition(), this).getAge()> 10){
            breeding(world.whoIs(this.position, this));
           this.setPosition(oldPosition);

        }
        /*
	else if(world->getElementInMap(this->getPozycja()) != nullptr && world->getElementInMap(this->getPozycja())->getModel() == this->getModel() && this->getAge() > 10 && world->getElementInMap(this->getPozycja())->getAge() > 10){

            Breeding(world->getElementInMap(this->getPozycja()));
            this->SetPozycja(getOldPosition().getX(), getOldPosition().getY());
        }
        */

        plusAge();

    }

    @Override
    public void Kolizja(Organizm org) {
        if (this.getSila()>org.getSila()) {
            world.deleteOrganizm(org);
            if (org.getSila() > 0) {
                //String result = this.getName()+" killed "+org.getName()+" on position " + this.getPosition().getStringPosition()+".";
                world.addInformation(this.getName()+" killed "+org.getName()+" on position "
                        + this.getPosition().getStringPosition()+".\n");

            }
        }
	else if (this.getSila() < org.getSila()) {

            world.addInformation(org.getName()+" killed "+this.getName()+" on position "
                    + this.getPosition().getStringPosition()+".\n");
            world.deleteOrganizm(this);
        }
	else if (this.getAge() > org.getAge()) {
            world.addInformation(this.getName()+" killed "+org.getName()+" on position "
                    + this.getPosition().getStringPosition()+".\n");
            world.deleteOrganizm(org);

        }
	else {
            world.addInformation(org.getName()+" killed "+this.getName()+" on position "
                    + this.getPosition().getStringPosition()+".\n");
            world.deleteOrganizm(this);
        }

    }

    @Override
    public String getName() {
        return null;
    }

    @Override
    public Image getImage() {
        return null;
    }

    public void breeding(Organizm org){
        boolean left = true;
        boolean right = true;
        boolean up = true;
        boolean down = true;
        if (this.getPosition().getX() -2 < 0) {
            left = false;
        }
        else if (0 + this.getPosition().getX() + 2> 0 + world.getX()) {
            right = false;
        }
        if (this.getPosition().getY() - 2 < 0) {
            up = false;
        }
        else if (this.getPosition().getY()+ 2> 0 + world.getY()) {
            down = false;
        }
        Point newPosition = new Point(1,1);
        if (world.whoIs(new Point(position.getX()+1, position.getY())) == null && right == true) {
            newPosition.setX(position.getX() + 1);
            newPosition.setY(position.getY());
        }
	else if (world.whoIs(new Point( position.getX() - 1, position.getY() )) == null && left == true) {
            newPosition.setX(position.getX() - 1);
            newPosition.setY(position.getY());
        }
        else if (world.whoIs(new Point( position.getX(), position.getY()-1 )) == null && up == true) {
            newPosition.setX(position.getX());
            newPosition.setY(position.getY()-1);
        }
        else if (world.whoIs(new Point( position.getX() , position.getY()+1 )) == null && down == true) {
            newPosition.setX(position.getX() );
            newPosition.setY(position.getY()+1);
        }
	else {
            return;
        }
        world.createOrganizm(getTyp(), newPosition);
    world.addInformation("A new "+this.getName()+" was born in the position "+newPosition.getStringPosition()+"\n");

    }

    }


